runtime/coverage_test.go

```diff
@@ -658,6 +659,295 @@ func TestCoverageReportAddLineHitForNonInspectedProgram(t *testing.T) {
+	require.NoError(t, err)
+
+	expected := `
+	  {
+	    "coverage": {
+	      "S.AnswerScript": {
+	        "line_hits": {
+	          "3": 1,
+	          "4": 1,
+	          "5": 0,
+	          "7": 0
+	        },
+	        "missed_lines": [5, 7],
+	        "statements": 4,
+	        "percentage": "50.0%"
+	      }
+	    }
+	  }
+	`
+	require.JSONEq(t, expected, string(actual))
+
+	assert.Equal(t, "50.0%", coverageReport.Percentage())
+}
+
+func TestCoverageReportCoveredStatementsPercentage(t *testing.T) {
+
+	t.Parallel()
+
+	script := []byte(`
+	  pub fun answer(): Int {
+	    var i = 0
+	    while i < 42 {
+	      i = i + 1
+	    }
+	    return i
+	  }
+	`)
+
+	program, err := parser.ParseProgram(nil, script, parser.Config{})
+	require.NoError(t, err)
+
+	coverageReport := NewCoverageReport()
+
+	location := common.StringLocation("AnswerScript")
+	coverageReport.InspectProgram(location, program)
+	coverageReport.AddLineHit(location, 3)
+	coverageReport.AddLineHit(location, 4)
+	coverageReport.AddLineHit(location, 5)
+
+	actual, err := json.Marshal(coverageReport)
+	require.NoError(t, err)
+
+	expected := `
+	  {
+	    "coverage": {
+	      "S.AnswerScript": {
+	        "line_hits": {
+	          "3": 1,
+	          "4": 1,
+	          "5": 1,
+	          "7": 0
+	        },
+	        "missed_lines": [7],
+	        "statements": 4,
+	        "percentage": "75.0%"
+	      }
+	    }
+	  }
+	`
+	require.JSONEq(t, expected, string(actual))
+
+	assert.Equal(
+		t,
+		"Coverage: 75.0% of statements",
+		coverageReport.CoveredStatementsPercentage(),
+	)
+}
+
+func TestCoverageReportDiff(t *testing.T) {
+
+	t.Parallel()
+
+	script := []byte(`
+	  pub fun answer(): Int {
+	    var i = 0
+	    while i < 42 {
+	      i = i + 1
+	    }
+	    return i
+	  }
+	`)
+
+	program, err := parser.ParseProgram(nil, script, parser.Config{})
+	require.NoError(t, err)
+
+	coverageReport := NewCoverageReport()
+
+	location := common.StringLocation("AnswerScript")
+	coverageReport.InspectProgram(location, program)
+	coverageReport.AddLineHit(location, 3)
+	coverageReport.AddLineHit(location, 4)
+
+	summary := coverageReport.Summary()
+
+	actual, err := json.Marshal(summary)
+	require.NoError(t, err)
+
+	expected := `
+	  {
+	    "coverage": "50.0%",
+	    "hits": 2,
+	    "locations": 1,
+	    "misses": 2,
+	    "statements": 4
+	  }
+	`
+	require.JSONEq(t, expected, string(actual))
+
+	otherCoverageReport := NewCoverageReport()
+	otherCoverageReport.InspectProgram(location, program)
+	otherCoverageReport.AddLineHit(location, 3)
+	otherCoverageReport.AddLineHit(location, 4)
+	otherCoverageReport.AddLineHit(location, 5)
+	otherCoverageReport.AddLineHit(location, 5)
+	otherCoverageReport.AddLineHit(location, 7)
+
+	diff := coverageReport.Diff(*otherCoverageReport)
+
+	actual, err = json.Marshal(diff)
+	require.NoError(t, err)
+
+	expected = `
+	  {
+	    "coverage": "100.0%",
+	    "hits": 2,
+	    "locations": 0,
+	    "misses": -2,
```

Comment:
Note: The negative value here represents the incremental diff, in comparison to the previous coverage report.